src/main/java/org/openrewrite/java/dependencies/RelocatedDependencyCheck.java

```diff
@@ -0,0 +1,283 @@
+import org.openrewrite.java.tree.JavaSourceFile;
+import org.openrewrite.marker.SearchResult;
+import org.openrewrite.maven.MavenIsoVisitor;
+import org.openrewrite.xml.XPathMatcher;
+import org.openrewrite.xml.tree.Xml;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class RelocatedDependencyCheck extends ScanningRecipe<RelocatedDependencyCheck.Accumulator> {
+    @Override
+    public String getDisplayName() {
+        return "Find relocated dependencies";
+    }
+
+    @Override
+    public String getDescription() {
+        //language=markdown
+        return "Find Maven and Gradle dependencies and Maven plugins that have relocated to a new `groupId` or `artifactId`. " +
+               "Relocation information comes from the [oga-maven-plugin](https://github.com/jonathanlermitage/oga-maven-plugin/) " +
+               "maintained by Jonathan Lermitage, Filipe Roque and others.";
+    }
+
+    @Value
+    public static class Accumulator {
+        Map<GroupArtifact, Relocation> migrations;
+    }
+
+    @Value
+    static class GroupArtifact {
+        String groupId;
+        @Nullable
+        String artifactId;
+    }
+
+    @Value
+    static class Relocation {
+        GroupArtifact to;
+        String context;
+    }
+
+    @Override
+    public Accumulator getInitialValue(ExecutionContext ctx) {
+        try {
+            MappingIterator<Migration> objectMappingIterator = new CsvMapper()
+                    .readerWithSchemaFor(Migration.class)
+                    .readValues(RelocatedDependencyCheck.class.getResourceAsStream("/migrations.csv"));
+            Map<GroupArtifact, Relocation> migrations = new HashMap<>();
+            while (objectMappingIterator.hasNext()) {
+                Migration def = objectMappingIterator.next();
+                GroupArtifact oldGav = new GroupArtifact(def.getOldGroupId(), StringUtils.isBlank(def.getOldArtifactId()) ? null : def.getOldArtifactId());
+                GroupArtifact newGav = new GroupArtifact(def.getNewGroupId(), StringUtils.isBlank(def.getNewArtifactId()) ? null : def.getNewArtifactId());
+                migrations.put(oldGav, new Relocation(newGav, StringUtils.isBlank(def.getContext()) ? null : def.getContext()));
+            }
+            return new Accumulator(migrations);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getScanner(Accumulator acc) {
+        return Preconditions.or(new IsBuildGradle<>(), new MavenIsoVisitor<ExecutionContext>() {
+            @Override
+            public @Nullable Xml visit(@Nullable Tree tree, ExecutionContext executionContext) {
+                return null; // Any Maven pom.xml file
+            }
+        });
```

Comment:
At the moment, we're not doing anything in the Scan phase, so we could probably just use `TreeVisitor.noop()` here.

Reply:
Thanks! That helps clear things out quite a bit.